File: CMT120/p_c.py
# Exercise 1
def reduceFraction(num, den):
    i = 2
    while i <= min(num, den):
        if num % i == 0 and den % i == 0:
            num = num / i
            den = den / i
        else:
            i = i + 1
    res = (num, den)
    return res


# Exercise 2
def isMagicDate(day, month, year):
    num = day * month
    last = year % 100
    if num == last:
        return True
    else:
        return False


# Exercise 3
def sublist(l):
    res = [[]]
    # i,j points the start and end of the list
    for i in range(0, len(l)):
        for j in range(i + 1, len(l) + 1):
            res.append(l[i:j])
    return res


# Exercise 4
def pigLatin(word):
    # mark whether the word begin with a uppercase
    capital_mark = word[0].isupper()
    # change the word to lowercase
    res = word.lower()
    # store the word last char, in case it is a punctuation
    last = word[-1:]
    last_mark = False
    if not last.isalpha():
        res = word[0:-1]
        last_mark = True
    if is_consonant(res):
        while (is_consonant(res)):
            res = res + res[0]
            res = res[1:]
        res = res + 'ay'
    else:
        res = res + 'way'
    if capital_mark:
        res = res.capitalize()
    if last_mark:
        res = res + last

    return res


# judge whether the word begin with consonant
def is_consonant(word):
    word = word.lower()
    res = word[0] in 'aeiou'
    if res:
        return False
    else:
        return True


# Exercise 5
def morseCode(message):
    morse = {
        'A': '.-',
        'B': '-...',
        'C': '-.-.',
        'D': '-..',
        'E': '.',
        'F': '..-.',
        'G': '--.',
        'H': '....',
        'I': '..',
        'J': '.---',
        'K': '-.-',
        'L': '.-..',
        'M': '--',
        'N': '-.',
        'O': '---',
        'P': '.--.',
        'Q': '--.-',
        'R': '.-.',
        'S': '...',
        'T': '-',
        'U': '..-',
        'V': '...-',
        'W': '.--',
        'X': '-..-',
        'Y': '-.--',
        'Z': '--..',
        '1': '.----',
        '2': '..---',
        '3': '...--',
        '4': '....-',
        '5': '.....',
        '6': '-....',
        '7': '--...',
        '8': '---..',
        '9': '----.',
        '10': '-----'
    }
    res = ''
    msg = message.upper()
    for i in range(0, len(msg)):
        # msg can only contain key as the char
        if msg[i].isalnum():
            res = res + morse[msg[i]] + ' '
    res = res[:-1]
    return res


# Exercise 6
def int2Text(num):
    ones = {
        '1': 'one', '2': 'two', '3': 'three', '4': 'four', '5': 'five',
        '6': 'six', '7': 'seven', '8': 'eight', '9': 'nine', '10': 'ten', '11': 'eleven',
        '12': 'twelve', '13': 'thirteen', '14': 'fourteen', '15': 'fifteen', '16': 'sixteen',
        '17': 'seventeen', '18': 'eighteen', '19': 'nineteen'
    }
    tens = {
        '2': 'twenty', '3': 'thirty', '4': 'forty', '5': 'fifty', '6': 'sixty', '7': 'seventy',
        '8': 'eighty', '9': 'ninety'
    }
    res = ''
    if num >= 100:
        hun = num // 100
        hun_str = str(hun)
        res = ones[hun_str] + ' ' + 'hundred' + ' '
        # refresh the num to the new number which hundred had been deleted
        num = num % 100
    if num >= 20:
        ten = num // 10
        ten_str = str(ten)
        res = res + tens[ten_str] + ' '
        num = num % 10
    if num > 0:
        num_str = str(num)
        res = res + ones[num_str] + ' '
    res = res[:-1]
    return res


# Exercise 7
def missingComment(filename):
    f = open(filename)
    res = []
    list = f.read().split('\n')
    if is_function(list[0]):
        res.append(function_name(list[0]))
    for i in range(1, len(list)):
        if is_function(list[i]):
            if not is_comment(list[i - 1]):
                res.append(function_name(list[i]))
    return res


# whether a string is format as a comment
def is_comment(str):
    if str == None or len(str) < 1:
        return False
    if str[0] == '#':
        return True
    else:
        return False


# whether a string is format as a function
def is_function(str):
    if str == None or len(str) < 1:
        return False
    if str[0:4] == 'def ':
        return True
    else:
        return False


# deal withe the function line to get function name
def function_name(str):
    str = str[4:]
    i = str.find('(')
    str = str[:i]
    return str


# Exercise 8
def consistentLineLength(filename, length):
    f = open(filename)
    str = ''
    for line in f.readlines():
        line = line.strip()
        str = str + line + ' '
    str = str[:-1]
    # convert the str into list, each element is a single word
    str_list = str.split()
    res = []
    i = 0
    while i <= len(str_list) - 1:
        s = ''
        # use tempLength to mark whether the sentence is under length
        temp_length = length
        while i <= len(str_list) - 1:
            if temp_length - len(str_list[i]) >= 0:
                s = s + str_list[i] + ' '
                temp_length = temp_length - len(str_list[i]) - 1
                i = i + 1
            else:
                break
        res.append(s[:-1])
    return res


# Exercise 9
def knight(start, end, moves):
    # use a grid to represent the board, so map the col to index by dictionary
    col = {'a': 0, 'b': 1, 'c': 2, 'd': 3, 'e': 4, 'f': 5, 'g': 6, 'h': 7}
    # convert the input parameter to grid's index
    start_row = int(start[1]) - 1
    start_col = col[start[0]]
    end_row = int(end[1]) - 1
    end_col = col[end[0]]
    # array represents the board, each element is 0
    arr = [[0 for i in range(0, 8)] for i in range(0, 8)]
    # change the element of target position to 1 !!!!!!!!
    arr[end_row][end_col] = 1
    res = knight_dfs(arr, start_row, start_col, moves)
    return res


'''
    inspired by the algorithm: depth first search,
    move every possible direction to judge whether it's on target position,
    arr ： a 8*8 grid,
    i,j ：the index of arr,
    n ：the given moves, limiting the depth of this function
'''


def knight_dfs(arr, i, j, n):
    # if it has moved maximum length, return false
    if n < 0:
        return False
    # if it move to the position beyond arr，return false
    if i < 0 or i > 7 or j < 0 or j > 7:
        return False
    # if it is in the target position, return true
    if arr[i][j] == 1:
        return True
    return (knight_dfs(arr, i + 1, j + 2, n - 1)
            or knight_dfs(arr, i + 1, j - 2, n - 1)
            or knight_dfs(arr, i - 1, j + 2, n - 1)
            or knight_dfs(arr, i - 1, j - 2, n - 1)
            or knight_dfs(arr, i + 2, j + 1, n - 1)
            or knight_dfs(arr, i + 2, j - 1, n - 1)
            or knight_dfs(arr, i - 2, j + 1, n - 1)
            or knight_dfs(arr, i - 2, j - 1, n - 1))


# Exercise 10
def warOfSpecies(environment):
    x = len(environment)
    y = len(environment[0])
    res = [['' for i in range(y)] for i in range(x)]
    result = ['' for i in range(x)]
    # edit_arr: modified array (given array added elements '1' at boundary )
    edit_arr = [['1' for i in range(y + 2)] for i in range(x + 2)]
    for i in range(1, x + 1):
        for j in range(1, y + 1):
            edit_arr[i][j] = environment[i - 1][j - 1]
    # loop each element in grid and change it
    for i in range(1, x + 1):
        for j in range(1, y + 1):
            char = edit_arr[i][j]
            dict = surround(edit_arr, i, j)
            res[i - 1][j - 1] = change(char, dict)
    # convert res to target format
    for i in range(0, x):
        for j in range(0, y):
            result[i] += res[i][j]
    return result


'''
    the function use the rule change the current species to the target one
    dict: is a dictionary
        key:surround species
        value:the number of key species
    char: current species
'''


def change(char, dict):
    if char == '.':
        if dict['X'] >= 2 or dict["O"] >= 2:
            if dict['X'] == dict['O']:
                return '.'
            if dict['X'] > dict['O']:
                return 'X'
            if dict['X'] < dict['O']:
                return 'O'
        else:
            return '.'
    if char == 'O':
        if dict['X'] + dict['O'] > 6 or dict['O'] < 3 or dict['O'] < dict['X']:
            return '.'
        else:
            return 'O'
    if char == 'X':
        if dict['X'] + dict['O'] > 6 or dict['X'] < 3 or dict['O'] > dict['X']:
            return '.'
        else:
            return 'X'


'''
    use dictionary sum to mark the number of species !!
    arr: the modified array (given array added elements '1' at boundary )
    i,j: current index
    return a dictionary with surround species and numbers
'''


def surround(list, i, j):
    sum = {'X': 0, 'O': 0, '.': 0, '1': 0}
    sum[list[i - 1][j - 1]] += 1
    sum[list[i - 1][j]] += 1
    sum[list[i - 1][j + 1]] += 1
    sum[list[i][j - 1]] += 1
    sum[list[i][j + 1]] += 1
    sum[list[i + 1][j - 1]] += 1
    sum[list[i + 1][j]] += 1
    sum[list[i + 1][j + 1]] += 1
    return sum

File: CMT120/p_c.js
// Exercise 1
function reduceFraction(num, den) {
    let min = Math.min(num, den);
    let div;
    let ans = new Array(2);
    //use for loop find the biggest div
    for (let i = 1; i <= min; i++) {
        if (num % i === 0 && den % i === 0) {
            div = i;
        }
    }
    ans[0] = num / div;
    ans[1] = den / div;
    return ans;
}

// Exercise 2
function isMagicDate(day, month, year) {
    let num = day * month;
    let last = year % 100;
    if (num === last) {
        return true
    } else
        return false;
}

// Exercise 3
function sublist(l) {
    let ans = [[]];
    let len = l.length;
    for (let i = 0; i < len; i++) {
        for (let j = i + 1; j <= len; j++) {
            let sub = l.slice(i, j);
            ans.push(sub);
        }
    }
    return ans;
}

// Exercise 4
function pigLatin(word) {
    let res = word.toLowerCase();
    // use a variable to mark the initial case: true/false
    let markCapital = isUppercase(word);
    // if the word end with punctuation mark, use a variable to store it, and delete it
    let last = '';
    if (hasPunctuation(word)) {
        last = last + word[word.length - 1];
        res = res.slice(0, word.length - 1);
    }
    //situation1: the word begins with a consonant
    if (isConsonant(res)) {
        while (isConsonant(res)) {
            res = res + res[0];
            res = res.slice(1);
        }
        res = res + 'ay';
    }
    //situation2: add 'way' in the end
    else {
        res = res + 'way';
    }
    if (markCapital) {
        res = res.charAt(0).toUpperCase() + res.slice(1);
    }
    res = res + last;
    return res;
}

//determine if the word begin with a consonant
function isConsonant(word) {
    let newWord = word.toLowerCase();
    if (newWord[0] === 'a' || newWord[0] === 'e' || newWord[0] === 'i' || newWord[0] === 'o' || newWord[0] === 'u') {
        return false;
    }
    return true;
}

//determine if the word begin with a uppercase
function isUppercase(word) {
    let newWord = word.toLowerCase();
    if (word == newWord) {
        return false;
    }
    return true;
}

//determine if the word end with a punctuation
function hasPunctuation(word) {
    let regChar = /[a-zA-Z]/;
    let l = word.charAt(word.length - 1);
    let res = regChar.test(l);
    if (res == true) {
        return false;
    }
    return true;
}

// Exercise 5
function morseCode(message) {
    const morse = {
        'A': '.-',
        'B': '-...',
        'C': '-.-.',
        'D': '-..',
        'E': '.',
        'F': '..-.',
        'G': '--.',
        'H': '....',
        'I': '..',
        'J': '.---',
        'K': '-.-',
        'L': '.-..',
        'M': '--',
        'N': '-.',
        'O': '---',
        'P': '.--.',
        'Q': '--.-',
        'R': '.-.',
        'S': '...',
        'T': '-',
        'U': '..-',
        'V': '...-',
        'W': '.--',
        'X': '-..-',
        'Y': '-.--',
        'Z': '--..',
        '1': '.----',
        '2': '..---',
        '3': '...--',
        '4': '....-',
        '5': '.....',
        '6': '-....',
        '7': '--...',
        '8': '---..',
        '9': '----.',
        '10': '-----'
    };
    let res = '';
    message = message.toUpperCase();
    let str = '';
    // str can only contain key as the char
    for (let i = 0; i <= message.length - 1; i++) {
        let char = message.charAt(i);
        if (isKey(char)) {
            str = str + char;
        }
    }
    for (let i = 0; i <= str.length - 1; i++) {
        res = res + morse[str.charAt(i)] + ' ';
    }
    res = res.slice(0, res.length - 1);
    return res;
}

//this function can filtering illegal characters
function isKey(char) {
    let regChar = /[A-Z0-9]/;
    let res = regChar.test(char);
    if (res == true) {
        return true;
    }
    return false;
}

// Exercise 6
function int2Text(num) {
    // dictionary ones maps the number<20
    const ones = {
        '1': 'one', '2': 'two', '3': 'three', '4': 'four', '5': 'five',
        '6': 'six', '7': 'seven', '8': 'eight', '9': 'nine', '10': 'ten', '11': 'eleven',
        '12': 'twelve', '13': 'thirteen', '14': 'fourteen', '15': 'fifteen', '16': 'sixteen',
        '17': 'seventeen', '18': 'eighteen', '19': 'nineteen'
    };
    //dictionary tens maps the decade of a number
    const tens = {
        '2': 'twenty', '3': 'thirty', '4': 'forty', '5': 'fifty', '6': 'sixty', '7': 'seventy',
        '8': 'eighty', '9': 'ninety'
    };
    let res = new String();
    if (num >= 100) {
        //variable hun store the hundred of the input number
        let hun = parseInt(num / 100);
        res = ones[hun] + ' ' + 'hundred' + ' ';
        //refresh the num to the new number which hundred had been deleted
        num = num % 100;
    }
    if (num >= 20) {
        let ten = parseInt(num / 10);
        res = res + tens[ten] + ' ';
        num = num % 10;
    }
    if (num > 0) {
        res = res + ones[num] + ' ';
    }
    //every res has been added '(blank)', so remove the last in the end
    res = res.slice(0, res.length - 1);
    return res;
}


// Exercise 7
function missingComment(filename) {
    let res = new Array();
    const fs = require("fs");
    const data = fs.readFileSync(filename, "utf8");
    //convert the data to array, elements is separated by line
    let arr = data.toString().split("\n");
    let len = arr.length;
    //if the fist line is a function, add it to result
    if (isFunction(arr[0])) {
        res.push(functionName(arr[0]));
    }
    //using a loop to find a function, then judge whether the previous line is a comment
    for (let i = 1; i <= len - 1; i++) {
        if (isFunction(arr[i])) {
            if (!(isComment(arr[i - 1]))) {
                res.push(functionName(arr[i]));
            }
        }
    }
    return res;
}

//whether a string is format as a comment
function isComment(str) {
    let reg = /\/\/.+/;
    let res = reg.test(str);
    return res;
}

//whether a string is format as a function
function isFunction(str) {
    let reg = /function\s.+/;
    let res = reg.test(str);
    return res;
}

//deal withe the function line to get function name
function functionName(str) {
    let sub = str.slice(9);
    let reg = /\s*\(.*/;
    let res = sub.replace(reg, '');
    return res;
}

// Exercise 8
function consistentLineLength(filename, length) {
    let res = new Array();
    const fs = require("fs");
    const data = fs.readFileSync(filename, "utf8");
    let reg = /[\s\n]/;
    //convert the data into arr, each element is a single word
    let arr = data.toString().split(reg);
    //delete with the '' in arr
    arr = deleteNull(arr);
    let i = 0;
    while (i <= arr.length - 1) {
        let s = '';
        //use tempLength to mark whether the sentence is under length
        let tmpLength = length;
        while (i <= arr.length - 1) {
            if (tmpLength - arr[i].length >= 0) {
                s += arr[i] + ' ';
                tmpLength -= (arr[i].length + 1);
                i++;
            } else {
                break;
            }
        }
        res.push(s.slice(0, s.length - 1));
    }
    return res;
}

//delete '' in an array
function deleteNull(arr) {
    for (let i = arr.length - 1; i >= 0; i--) {
        if (arr[i] == '') {
            arr.splice(i, 1);
        }
    }
    return arr;
}

// Exercise 9
function knight(start, end, moves) {
    //use a grid to represent the board, so map the col to index by dictionary
    let col = {'a': 0, 'b': 1, 'c': 2, 'd': 3, 'e': 4, 'f': 5, 'g': 6, 'h': 7};
    //convert the input parameter to grid's index
    let startRow = (start.charAt(1)) - 1;
    let startCol = col[start.charAt(0)];
    let endRow = (end.charAt(1)) - 1;
    let endCol = col[end.charAt(0)];
    //array represents the board, each element is 0
    let arr = new Array();
    for (let i = 0; i <= 7; i++) {
        arr[i] = new Array();
        for (let j = 0; j <= 7; j++) {
            arr[i][j] = 0;
        }
    }
    //change the element of target position to -1 !!!!!!!!
    arr[endRow][endCol] = -1;
    //dfs
    let res = dfs(arr, startRow, startCol, moves);
    return res;
}

/*
    inspired by the algorithm: depth first search,
    move every possible direction to judge whether it's on target position,
    arr is a 8*8 grid,
    i,j is the index of arr,
    n is the given moves, limiting the depth of this function
 */
function dfs(arr, i, j, n) {
    // if it has moved maximum length, return false
    if (n < 0) {
        return false;
    }
    // if it move to the position beyond arr，return false
    if (i < 0 || j < 0 || i > 7 || j > 7) {
        return false;
    }
    // if it is in the target position, return true
    if (arr[i][j] == -1) {
        return true;
    }
    // the function start and recursion here, it will go every path with limited moves
    return dfs(arr, i + 2, j + 1, n - 1) ||
        dfs(arr, i + 2, j - 1, n - 1) ||
        dfs(arr, i - 2, j + 1, n - 1) ||
        dfs(arr, i - 2, j - 1, n - 1) ||
        dfs(arr, i + 1, j + 2, n - 1) ||
        dfs(arr, i + 1, j - 2, n - 1) ||
        dfs(arr, i - 1, j + 2, n - 1) ||
        dfs(arr, i - 1, j - 2, n - 1);

}

// Exercise 10
function warOfSpecies(environment) {
    let result = new Array();
    let res = new Array();
    for (let i = 0; i <= environment.length - 1; i++) {
        res[i] = new Array();
        for (let j = 0; j <= environment[0].length - 1; j++) {
            res[i][j] = '';
        }
    }
    //editArr: modified array (given array added elements '1' at boundary )
    let editArr = new Array();
    for (let i = 0; i <= environment.length + 1; i++) {
        editArr[i] = new Array();
        for (let j = 0; j <= environment[0].length + 1; j++) {
            if (i == 0 || i == environment.length + 1 || j == 0 || j == environment[0].length + 1) {
                editArr[i][j] = '1';
            } else {
                editArr[i][j] = environment[i - 1][j - 1];
            }

        }
    }
    //loop each element in grid and change it
    for (let i = 1; i <= environment.length; i++) {
        for (let j = 1; j <= environment[0].length; j++) {
            let char = editArr[i][j];
            let dict = surround(editArr, i, j);
            res[i - 1][j - 1] = change(char, dict);
        }
    }
    // convert res to target format
    for (let i = 0; i <= environment.length - 1; i++) {
        result[i] = new String();
        for (let j = 0; j <= environment[0].length - 1; j++) {
            result[i] += res[i][j];
        }
    }
    return result;
}

/*
    the function use the rule change the current species to the target one
    dict: is a dictionary
        key:surround species
        value:the number of key species
    char: current species

*/
function change(char, dict) {
    if (char == '.') {
        if (dict['X'] >= 2 || dict["O"] >= 2) {
            if (dict['X'] == dict['O']) {
                return '.';
            }
            if (dict['X'] > dict['O']) {
                return 'X';
            }
            if (dict['X'] < dict['O']) {
                return 'O';
            }
        } else {
            return '.';
        }
    }
    if (char == 'O') {
        if (dict['X'] + dict['O'] > 6 || dict['O'] < 3 || dict['O'] < dict['X']) {
            return '.';
        } else {
            return 'O';
        }
    }
    if (char == 'X') {
        if (dict['X'] + dict['O'] > 6 || dict['X'] < 3 || dict['X'] < dict['O']) {
            return '.';
        } else {
            return 'X';
        }
    }
}

/*
    use dictionary sum to mark the number of species !!
    arr: the modified array (given array added elements '1' at boundary )
    i,j: current index
    return a dictionary with surround species and numbers
 */
function surround(arr, i, j) {
    let sum = {'X': 0, 'O': 0, '.': 0, '1': 0};
    //if it surround position is the key of sum, refresh it value by +1;
    sum[arr[i - 1][j - 1]] += 1;
    sum[arr[i - 1][j]] += 1;
    sum[arr[i - 1][j + 1]] += 1;
    sum[arr[i][j - 1]] += 1;
    sum[arr[i][j + 1]] += 1;
    sum[arr[i + 1][j - 1]] += 1;
    sum[arr[i + 1][j]] += 1;
    sum[arr[i + 1][j + 1]] += 1;
    return sum;
}


module.exports = {
    reduceFraction: reduceFraction,
    isMagicDate: isMagicDate,
    sublist: sublist,
    pigLatin: pigLatin,
    morseCode: morseCode,
    int2Text: int2Text,
    missingComment: missingComment,
    consistentLineLength: consistentLineLength,
    knight: knight,
    warOfSpecies: warOfSpecies
}